### DISCUSSION TITLE
How do I define a receiver in the TracerProviderConfigurer

### DISCUSSION THREAD

**Discussion by Cavvar** [open-telemetry/opentelemetry-java#3584] [Q&A] [answered]
Hello everyone!

I'm writing a java-agent-extension that implements a  `CustomSpanProcessor` that receives an `OtlpGrpcSpanExporter` in the constructor. The `CustomSpanProcessor` sets attributes during `onStart()` with the writable span but also has to set attributes during the `onEnd()` hook. Since the hook is only readable, the workaround is to convert the `ReadableSpan` into a writable format and then send to a custom exporter. Ideally a hook like `beforeEnd()` (as mentioned in [Issue#1089](https://github.com/open-telemetry/opentelemetry-specification/issues/1089)) exists.

The configurer roughly looks like the following:
```java
import com.google.auto.service.AutoService;
import io.opentelemetry.exporters.otlp.OtlpGrpcSpanExporter;
import io.opentelemetry.sdk.autoconfigure.spi.SdkTracerProviderConfigurer;
import io.opentelemetry.sdk.trace.SdkTracerProviderBuilder;

@AutoService(SdkTracerProviderConfigurer.class)
public class DemoSdkTracerProviderConfigurer implements SdkTracerProviderConfigurer {

    @Override
    public void configure(SdkTracerProviderBuilder sdkTracerProviderBuilder) {
        sdkTracerProviderBuilder
                .addSpanProcessor(new CustomSpanProcessor(OtlpGrpcSpanExporter.getDefault()));
    }
}
```

My test application is made with Quarkus. I have added the java-agent and my agent-extension into the docker image and extended the `JAVA_OPTIONS` parameter.
```docker
# Configure the JAVA_OPTIONS, you can add -XshowSettings:vm to also display the heap size.
ENV JAVA_OPTIONS="-javaagent:/agents/opentelemetry-javaagent-all.jar \
                  -Dotel.traces.exporter=otlp \
                  -Dotel.exporter.otlp.endpoint=http://jaeger:14250 \
                  -Dotel.resource.attributes=service.name=test-service \
                  -Dotel.metrics.exporter=none \
                  -Dotel.javaagent.extensions=/agents/javaagent.extension-1.0.0-SNAPSHOT.jar \
                  -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager"
```

When running the application through my docker compose I get an error from opentelemetry.
<img width="1803" alt="ErrorConfigNoReceiver" src="https://user-images.githubusercontent.com/3383908/132300678-cd06fdfb-51e6-4a5f-9af0-3579eab80f35.png">

Is there a way to define a receiver with the setup I have?

**Comment by Cavvar** [comment, marked as answer]
I think I answered the question myself.

Instead of using a `OtlpGrpcSpanExporter` I am using a `JaegerGrpcSpanExporter` now. For that I had to set the endpoint or channel. This caused the Jaeger Backend to display duplicate spans. I fixed that by simple setting the env-variable `-Dotel.traces.exporter=none`.

If anyone needs any clarification, feel free to ask!

**Reply by jkwatson** [reply]
Glad you got a solution. What are you trying to do here, btw? Export in-flight spans?

**Reply by Cavvar** [reply]
Not sure what in-flight spans are exactly. I am trying to explain my problem as clearly as possible.

I want to set attributes of a span during its start and after it has ended. As you know, spans during `onEnd()` are immutable. So the workaround I researched was to transform spans during `onEnd` into a readable format and then send them to an `SpanExporter`.

Ideally, if the exporter has been set to use jaeger (`Dotel.traces.exporter=jaeger`) I would want to be able to obtain that exporter through the `SdkTracerProviderConfigurer` or atleast inside the overriden `configure()` method. If that would be possible, then I could give my custom `SpanProcessor` that exporter and would avoid duplicate spans. The workaround right now for that would be to define `none` as mentioned in another [discussion](https://github.com/open-telemetry/opentelemetry-java/discussions/3033).

As I mentioned in my initial post, ideally a hook like `beforeEnd()` exists as discussed in [Issue#1089](https://github.com/open-telemetry/opentelemetry-specification/issues/1089). I'll might add to that issue that interests in that hook still exists.

I'm sorry if anything wasn't clear enough.

**Reply by Cavvar** [reply]
Based on the defined exporter I would want to decide which in the end I want to use. I have figured it out now by myself again I think.
I have set both metrics.exporter and traces.exporter to none and added my own JAVA_OPTIONS. Based on my custom properties I added a simple if check. Hope this clarified my intentions.

**Reply by jkwatson** [reply]
I think I finally (sorry) grok that you're asking for auto-configuration/SPI support for specifying an exporter via auto-configuration, and having it injected into your `SdkTracerProviderConfigurer` implementation.

I think this would start to introduce a tremendous amount of complexity in the SPI interfaces. In this case, I think you're probably best off just fully implementing what you need in your own implementation.

@anuraaga any thoughts?
